pkg/bff/organization.go

```diff
@@ -13,6 +15,145 @@ import (
+			return
+		}
+	}
+
+	// Create a new organization in the database
+	var org *models.Organization
+	if org, err = s.db.CreateOrganization(); err != nil {
+		log.Error().Err(err).Msg("could not create organization in database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create organization"))
+		return
+	}
+
+	// Update the organization with the request params
+	org.Name = params.Name
+	org.Domain = domain
+	if err = s.db.UpdateOrganization(org); err != nil {
+		log.Error().Err(err).Msg("could not update organization in database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create organization"))
+		return
+	}
+
+	// Assign the user to the organization
+	appdata.AddOrganization(org.Id)
+	if err = s.SaveAuth0AppMetadata(*user.ID, *appdata); err != nil {
+		log.Error().Err(err).Msg("could not update user app metadata")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not update user app metadata"))
+		return
+	}
+
+	// Build the response
+	out := &api.OrganizationReply{
+		ID:           org.Id,
+		Name:         org.Name,
+		Domain:       org.Domain,
+		CreatedAt:    org.Created,
+		RefreshToken: true,
+	}
+	c.JSON(http.StatusOK, out)
+}
+
+// ListOrganizations returns a list of organizations that the user is a member of. The
+// user must have the read:organizations permission to perform this action.
+func (s *Server) ListOrganizations(c *gin.Context) {
+	var (
+		err  error
+		user *management.User
+	)
+
+	// Fetch the user from the context
+	if user, err = auth.GetUserInfo(c); err != nil {
+		log.Error().Err(err).Msg("list organizations handler requires user info; expected middleware to return 401")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not identify user to list organizations"))
+		return
+	}
+
+	// Load the user app metadata to check their organization assignments
+	appdata := &auth.AppMetadata{}
+	if err = appdata.Load(user.AppMetadata); err != nil {
+		log.Error().Err(err).Msg("could not parse user app metadata")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not parse user app metadata"))
+		return
+	}
+
+	// Build the response
+	// TODO: Return a last login timestamp so the frontend can order by last used
+	out := make([]*api.OrganizationReply, 0)
+	for _, id := range appdata.GetOrganizations() {
+		if org, err := s.OrganizationFromID(id); err != nil {
+			log.Error().Err(err).Str("org_id", id).Msg("could not retrieve organization from database")
+		} else {
+			out = append(out, &api.OrganizationReply{
+				ID:        org.Id,
+				Name:      org.Name,
+				Domain:    org.Domain,
+				CreatedAt: org.Created,
+			})
+		}
+	}
+
+	c.JSON(http.StatusOK, out)
+}
+
+// Normalize a domain name for matching purposes.
+func normalizeDomain(domain string) string {
+	return strings.ToLower(strings.TrimRight(strings.TrimSpace(domain), "."))
```

Comment:
I'm thinking that we'll want to strip off any `http://` that's part of the domain as well, so we could perhaps optimistically parse the URL (but if it doesn't parse, then ignore it).

```suggestion
	if u, err := url.Parse(domain); err != nil {
		domain = u.Host()
	}    
	return strings.ToLower(strings.TrimRight(strings.TrimSpace(domain), "."))
```

We may also want to have a regular expression to validate the domain - something pretty permissive, but to prevent things like just `"foo"` or an email address. E.g. `^(((?!\-))(xn\-\-)?[a-z0-9\-_]{0,61}[a-z0-9]{1,1}\.)*(xn\-\-)?([a-z0-9\-]{1,61}|[a-z0-9\-]{1,30})\.[a-z]{2,}$` from https://stackoverflow.com/questions/10306690/what-is-a-regular-expression-which-will-match-a-valid-domain-name-without-a-subd -- though some tests would be good to make sure it does what we want.